source/slang/ir-specialize.cpp

```diff
@@ -1145,6 +1153,167 @@ struct SpecializationContext
+        IRBuilder*  builder,
+        IRType*     type)
+    {
+        // The "true" pointers and the pointer-like stdlib types are the easy cases.
+        if( auto ptrType = as<IRPtrTypeBase>(type) )
+        {
+            return ptrType->getValueType();
+        }
+        else if( auto ptrLikeType = as<IRPointerLikeType>(type) )
+        {
+            return ptrLikeType->getElementType();
+        }
+        //
+        // A more interesting case arises when we have a `BindExistentials<P<T>, ...>`
+        // where `P<T>` is a pointer(-like) type.
+        //
+        else if( auto bindExistentials = as<IRBindExistentialsType>(type) )
+        {
+            // We know that `BindExistentials` won't introduce its own
+            // existential type parameters, nor will any of the pointer(-like)
+            // type constructors `P`.
+            //
+            // Thus we know that the type that is pointed to should be
+            // the same as `BindExistentials<T, ...>`.
+            //
+            auto baseType = bindExistentials->getBaseType();
+            if( auto baseElementType = tryGetPointedToType(builder, baseType) )
+            {
+                UInt existentialArgCount = bindExistentials->getExistentialArgCount();
+                List<IRInst*> existentialArgs;
+                for( UInt ii = 0; ii < existentialArgCount; ++ii )
+                {
+                    existentialArgs.Add(bindExistentials->getExistentialArg(ii));
+                }
+                return builder->getBindExistentialsType(
+                    baseElementType,
+                    existentialArgCount,
+                    existentialArgs.Buffer());
+            }
+        }
+
+        // TODO: We may need to handle other cases here.
+
+        return nullptr;
+    }
+
+    void maybeSpecializeLoad(IRLoad* inst)
+    {
+        auto ptrArg = inst->ptr.get();
+
+        if( auto wrapInst = as<IRWrapExistential>(ptrArg) )
+        {
+            // We have an instruction of the form `load(wrapExistential(val, ...))`
+            //
+            auto val = wrapInst->getWrappedValue();
+
+            // We know what type we are expected to
+            // produce (which should be the pointed-to
+            // type for whatever the type of the
+            // `wrapExistential` is).
+            //
+            auto resultType = inst->getFullType();
+
+            IRBuilder builder;
+            builder.sharedBuilder = &sharedBuilderStorage;
+            builder.setInsertBefore(inst);
+
+            // We'd *like* to replace this instruction with
+            // `wrapExistential(load(val))` instead, since that
+            // will enable subsequent specializations.
+            //
+            // To do that, we need to be able to determine
+            // the type that `load(val)` should return.
+            //
+            auto elementType = tryGetPointedToType(&builder, val->getDataType());
+            if(!elementType)
```

Comment:
The idea of all the `maybeSimplify...()` operations is that they only simplify if all the operands/types have the appropriate forms, but it isn't an error when those conditions aren't met - it might just mean that one or more of the operands/types involved needs to be simplified itself, after which we might re-try this simplification and have it work.

We should only diagnose in a case where we are 100% certain that further simplification is needed, but impossible.